CalendarPicker/index.js

```diff
@@ -177,31 +177,111 @@ export default class CalendarPicker extends Component {
       onDateChange,
       enableDateChange,
       allowBackwardRangeSelect,
+      maxRangeDuration,
     } = this.props;
 
     if (!enableDateChange) {
       return;
     }
 
-    const date = moment({ year: currentYear, month: currentMonth, day, hour: 12 });
+    const date = moment({ year: currentYear, month: currentMonth, day, hour: 12 }),
+      startDate = selectedStartDate
+        ? moment(selectedStartDate)
+        : selectedStartDate,
+      endDate = selectedEndDate ? moment(selectedEndDate) : selectedEndDate;
 
-    if (allowRangeSelection && selectedStartDate && !selectedEndDate) {
-      if (date.isSameOrAfter(selectedStartDate, "day")) {
+    if (allowRangeSelection && startDate && !endDate) {
```

Comment:
This code block has become quite long and convoluted, making maintenance difficult.  Please compartmentalize the `allowBackwardRangeSelect` logic into a separate function and call that to process.  For example:
```
if (allowBackwardRangeSelect) {
  // Perform new backward range logic
} else {
  // Previous existing logic
}
```

Once you do that, you might find commonality between them and be able to re-use code.

Reply:
Just saw this.
I'll submit another PR for the abstraction.

Reply:
On second thought, abstracting the logic for reverse selection seems redundant.  Instead, I think that it should be a default part of the plugin logic as selecting two dates in any order to create a range seems like the normal user experience. I've closed my other PR.  If someone else wants to take up refactoring, I will leave the branch open.

Reply:
Maintenance is an ongoing consideration for this new logic. The original logic for range selection is much more direct than the added logic to support backward range selection.  The intent of range selection is lost in the nested if-else branches that were added.  It makes the code harder to follow for a particular path (e.g. forward selection jumps from this if, to an else, to another else, etc).  That results in brittle code when someone else fixes a bug in one path but accidentally alters the logic for a different path.

Reply:
Yeah, it sounds as though there needs to be a broader refactor of the range selection logic that allows bi-directional selection without a ton of complexity.  I am not sure if a simple abstraction will solve this - as allowBackwardRangeSelect will still be an ongoing dependency for any future maintenance consideration. I get the time in the future, I will revisit.